### ISSUE TITLE
Step 2: GPU Global Memory Coalescing

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11943]
Step 2: GPU Global Memory Coalescing<br>
<br>
https://ift.tt/ukCQYlB<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<figure></figure><hr>
<p>Looking at the GFLOPS from our kernel function and cuBLAS SGEMM (Figure 1), there are two main questions:</p>
<ol>
<li>Why are GFLOPS increasing with matrix sizes?</li>
<li>How can we improve the performance of our kernel function?</li>
</ol>
<figure><img src="https://0mean1sigma.com/content/images/2024/09/image-1.png"><figcaption><span>Figure 1: cuBLAS vs My Code</span></figcaption></figure><p>To answer these questions we need to understand the GPU hardware better.</p>
<h1>Modern GPU Architecture</h1>
<p>Figure 2 shows a high-level CUDA C programmer's view of a CUDA-capable GPU's architecture. There are <strong>four</strong> key features in this architecture:</p>
<ul>
<li>The GPU is organized into an array of highly threaded <strong>streaming multiprocessors (SMs)</strong>. </li>
<li>Each SM has several processing units called streaming processors or <strong>CUDA cores </strong>(shown as green tiles inside SMs) that share control logic.</li>
<li>The SMs also contain a different <strong>on-chip memory</strong> shared amongst the CUDA cores inside the SM. </li>
<li>GPU also has a much larger <strong>off-chip memory</strong>, the main component of which is the global memory or VRAM.</li>
</ul>
<figure><img src="https://0mean1sigma.com/content/images/2024/07/GPUarch.png"><figcaption><span>Figure 2: Architecture of a CUDA-capable GPU</span></figcaption></figure><p>There are two main takeaways from Figure 2:</p>
<ol>
<li>CUDA cores in different SMs can't interact with each other or access memory from other SMs.</li>
<li>As the global memory is off-chip (i.e., some distance away from the cores), it has <strong>long latency</strong> and <strong>low bandwidth</strong>.</li>
</ol>
<div>
<p>💡</p>
<p>Latency is the time it takes for data to travel from one point to another, usually measured in milliseconds (ms). Bandwidth is the maximum amount of data that can be sent and received in a given time, usually measured in Mbps or Gbps.</p>
</div>
<h1>Thread Blocks and SMs</h1>
<p>When a kernel is called, the CUDA runtime system launches a grid of threads that execute the same kernel code. These threads are assigned to SMs on a block-by-block basis, i.e., all threads in a block are simultaneously assigned to the same SM. For the matrix multiplication example, multiple blocks will likely get simultaneously assigned to the same SM. </p>
<figure><img src="https://0mean1sigma.com/content/images/2024/09/TB-SM.png"><figcaption><span>Figure 3: SMs and CUDA Thread Blocks</span></figcaption></figure><p>The example discussed in Figure 3 is quite small. In real-world problems, there are a lot more blocks, and to ensure that all blocks get executed, the runtime system maintains a list of blocks that did not get assigned to any SM and assigns these new blocks to SMs when previously assigned blocks complete execution. This block-by-block assignment of threads guarantees that threads in the same block are executed simultaneously on the same SM, which makes interaction between threads in the same block possible.</p>
<p>For a moment, it might look like an odd choice not to let threads in different blocks interact with each other. However, this feature allows different blocks to run independently in any order, resulting in<em> </em><strong>transparent scalability</strong> where the same code can run on different hardware with different execution resources. This, in turn, reduces the burden on software developers and ensures that with new generations of hardware, the application will speed up consistently without errors.</p>
<figure><img src="https://0mean1sigma.com/content/images/2024/07/Scalability.png"><figcaption><span>Figure 4: </span><a rel="noreferrer" href="https://www.google.com/books/edition/Programming_Massively_Parallel_Processor/7H9dEAAAQBAJ?hl=en&amp;gbpv=1&amp;dq=programming+massively+parallel+processors&amp;pg=PP1&amp;printsec=frontcover&amp;ref=0mean1sigma.com"><span>A Smaller GPU (on the left) can only run 2 blocks concurrently, while a larger GPU (on the right) can run 4 blocks concurrently</span></a><span>.</span></figcaption></figure><h1>Warps</h1>
<p>In the previous section, I explained that blocks can execute in any order relative to each other, but I did not say anything about how threads inside each block are executed. Conceptually, the programmer should assume that threads in a block can execute in any order, and the correctness of the algorithm should not depend on the order in which threads are executed. </p>
<p>Thread scheduling in CUDA GPUs is a hardware implementation concept that varies depending on the type of hardware used. In most implementations, once a block is assigned to an SM, it is divided into 32-thread units called <strong>warps</strong>. The knowledge of warps is useful for understanding and optimizing the performance of CUDA applications.</p>
<div>
<p>💡</p>
<p>Warp size can be different for future hardware generations!</p>
</div>
<p>Each warp consists of 32 consecutive threads. For 1D block, this is straightforward such that <code>threadIdx.x</code> is used to divide the threads into warps.</p>
<figure><img src="https://0mean1sigma.com/content/images/2024/07/1DWarp.png"><figcaption><span>Figure 5: Warps in 1D blocks</span></figcaption></figure><div>
<p>💡</p>
<p>If the number of threads in a block is not a multiple of 32, the last warp gets padded with inactive threads to fill up the remaining thread positions.</p>
</div>
<p>For the 2D block, threads are linearized using a row-major layout and partitioned into warps like the 1D case. For the 3D block, the <em>z</em> dimension is folded into the <em>y</em> dimension, and the resulting thread layout is treated as the 2D case.</p>
<figure><img src="https://0mean1sigma.com/content/images/2024/07/Warp.png"><figcaption><span>Figure 6: Warps in 2D and 3D blocks</span></figcaption></figure><p>An SM is designed to execute all threads in a warp following the SIMD model, i.e., at any instance in time, one instruction is fetched and executed for all threads in the warp. As one instruction is shared across multiple execution units, it allows for a small percentage of the hardware to be dedicated to control, and a large percentage is dedicated to increasing the arithmetic throughput (i.e., cores).</p>
<h1>Kernel Function</h1>
<p>There are three main steps to the running a program on the GPU:</p>
<ol>
<li>Copy data from host memory to device global memory.</li>
<li>Perform computations using the device cores and the data stored in device global memory.</li>
<li>Copy results from device global memory to host memory.</li>
</ol>
<p>We can't do much about steps 1 and 3 as we have to move data between host and device memory. To be honest this is computationally intensive but again, it's a one time thing and with large data, we can offset the cost with gains from solving the problem in parallel. So my focus here is to analyze the kernel function.</p>
<p>As a recap, the algorithm for parallel matrix multiplication involving matrix <em><strong>A</strong></em> of size (<em>M</em> x <em>K</em>) and <em><strong>B</strong></em> of size (<em>K</em> x <em>N</em>) is as follows:</p>
<ol>
<li>\(M \cdot N\) threads are generated on the GPU (one to compute each element of matrix \(C\) which is of size <em>M</em> x <em>N</em>).</li>
<li>Each thread:<ol>
<li>Retrieves a row of matrix \(A\) and a column of matrix \(B\) from the device memory. This results in the total of 2 x 4K Bytes being copied from device global memory.</li>
<li>Loops over the elements.</li>
<li>Multiplies the two numbers and add the result to the total, which is stored back in the device global memory. This results in the total of 4 Bytes being copied to device memory.</li>
</ol>
</li>
</ol>
<p>For all M x N threads, we are accessing M x N x (2 x 4K + 4) Bytes. If we have 4096 x 4096 matrices, the accesses from device global memory amount to 512 GB! I'm focusing so much on these memory accesses because global memory has <strong>long latency</strong> and <strong>low bandwidth</strong>, which is usually a major bottleneck for most applications.</p>
<h1>Answering the Questions</h1>
<h2>Why GFLOPS increase with increasing Matrix Size?</h2>
<p>From the knowledge of the GPU hardware (that we have acquired so far), it looks like the loss in performance for matrix multiplication with small matrix sizes (Figure 1) is mostly due to the global memory accesses. However, this does not explain why GFLOPS increase with the increase in the matrix size. I mean that with large matrices, the number of global memory accesses also increases, but the counter intuitively, the performance increases as well! GPUs can do this because the hardware is capable of <strong>latency tolerance</strong> or <strong>latency hiding</strong>.</p>
<p>There are usually more threads assigned to an SM than its cores. This is done so that GPUs can tolerate long-latency operations (like global memory accesses). With enough warps, SM can find a warp to execute while others are waiting for long-latency operations (like getting data from global memory). Filling the latency time of operations from some threads with work from others is called <strong>latency tolerance</strong> or <strong>latency hiding</strong>. The selection of warps ready for execution does not introduce any computational cost because GPU hardware is designed to facilitate <strong>zero-overhead thread scheduling</strong>.</p>
<div>
<p>💡</p>
<p>This context-switching can also be done with CPU threads. In CPUs, before switching to a different thread, the current execution state is moved from registers to memory and later loaded back to registers (from memory). This saving and restoring contents in a CPU can incur significant overhead.</p>
</div>
<div>
<p>💡</p>
<p>GPUs perform context switching differently, such that the execution state is held in place (inside thread registers). The ability to tolerate such long latency operations is why GPUs have smaller cache memories and control logic but way more registers. Another distinction from CPU is that GPU registers are not designed to store data but act as a staging area for threads.</p>
</div>
<p>That's why, for large matrices, more warps are available to hide the latency due to global memory accesses. There is a limit (set by CUDA) to the number of warps that can be assigned to an SM. However, it's not possible to assign an SM with the maximum number of warps that it supports because of constraints on execution resources (like on-chip memory) in an SM. The resources are dynamically partitioned such that SMs can execute many blocks with few threads or a few blocks with many threads. </p>
<p>For example, an Ampere A100 GPU can support 32 blocks per SM, 64 warps (2048 threads) per SM, and 1024 threads per block</p>
<p>So, if</p>
<ol><li>A grid is launched with 1024 threads in a block (maximum allowed)</li></ol>
<p><strong>Ans.</strong> Each SM can accommodate 2 blocks (with 2048 threads total, matching the maximum allowed per SM).</p>
<ol start="2"><li>A grid is launched with 512 threads in a block</li></ol>
<p><strong>Ans.</strong> Each SM can accommodate 4 blocks (with 2048 threads total, matching the maximum allowed per SM).</p>
<ol start="3"><li>A grid is launched with 256 threads in a block</li></ol>
<p><strong>Ans.</strong> Each SM can accommodate 8 blocks (with 2048 threads total, matching the maximum allowed per SM).</p>
<ol start="4"><li>A grid is launched with 64 threads in a block</li></ol>
<p><strong>Ans.</strong> Each SM can accommodate 32 blocks (with 2048 threads total, matching the maximum allowed per SM).</p>
<div>
<p>💡</p>
<p>The ratio of the number of warps assigned to an SM to the maximum number it supports is known as <b><strong>occupancy</strong></b>.</p>
</div>
<p>A negative situation might arise when the maximum number of threads allowed per block is not divisible by the block size. For example, we know that an Ampere A100 GPU can support 2048 threads per SM</p>
<p>So, if</p>
<ol><li>A grid is launched with 700 threads in a block</li></ol>
<p><strong>Ans.</strong> SM can hold only 2 blocks (totaling 1400 threads), and the remaining 648 thread slots are not utilized. The occupancy in this case is 1400 (assigned threads) / 2048 (maximum threads) = 68.35%.</p>
<div>
<p>💡</p>
<p>I did not consider the impact of other resources like memory. Automatic variables declared in a CUDA kernel are stored in registers, and that might not allow a large number of blocks to be allocated for execution to an SM.</p>
</div>
<h2>How to improve Kernel Performance?</h2>
<p>For large matrices, there will be enough threads assigned to each SM such that the hardware can get around the issue of global memory latency. To improve the kernel performance, we must makes efficiently accessing data from global memory a top priority.</p>
<p>To effectively utilize the global memory bandwidth, we must understand a few things about the architecture of the global memory. Global memory in CUDA devices is implemented using DRAM (Dynamic Random Access Memory) technology, which usually takes tens of nanoseconds to access a data byte. This starkly contrasts modern computing devices where the access speed is sub-nanosecond per byte. As the DRAM access speed is relatively slow, parallelism is used to increase the rate of data access (also known as memory access throughput).</p>
<p>Each time a DRAM location is accessed, a range of consecutive locations are also accessed in parallel. These consecutive location accesses and delivery are known as DRAM bursts. Current CUDA devices employ a technique that takes advantage of the fact that threads in a warp execute the same instruction at any given time. When all threads in a warp execute a load instruction, the hardware detects whether the memory accesses are consecutive or not. If they are consecutive, a lot of data can be transferred in parallel faster. In short, all threads in a warp must access consecutive global memory locations for optimum performance (this is known as coalesced access).</p>
<p>Figure 7 shows the pictorial analysis of the thread to element mapping. It's worth noting that the consecutive threads the warp are mapped to the columns of the output matrix. This means that each thread accesses the same column of matrix <em>B</em> and consecutive rows of matrix <em>A</em>. For any iteration <code>k</code>, threads in this warp access the elements of matrix <em>A</em>, which are <em>N</em> elements apart in the memory (remember that the 2D matrix is stored in a row-major layout in memory). As the threads in a warp are not accessing the consecutive elements in the memory, DRAM can not transfer these in parallel, hence requiring a separate load cycle for each element (Figure 8)! </p>
<figure><img src="https://0mean1sigma.com/content/images/2024/11/Step2_Figure7.gif"><figcaption><span>Figure 7: Uncoalesced thread to element mapping</span></figcaption></figure><figure><img src="https://0mean1sigma.com/content/images/2024/11/Step2_Figure8.gif"><figcaption><span>Figure 8: Uncoalesced memory accesses</span></figcaption></figure><hr>
<p>The solution to this problem is very simple. All we need to do is ensure that the threads in a warp are accessing consecutive elements of <em><strong>A</strong></em>. Figure 9 shows thread to element mapping for coalesced memory accesses. In this case, consecutive threads in the warp are responsible for computing elements along the rows of matrix <em>C</em>. This means that each thread accesses the same row of matrix <em>A</em> and consecutive columns of matrix <em>B</em>. For any iteration <code>k</code>, threads in this warp access the consecutive elements of matrix <em>B</em>. That means all the elements will be transferred in parallel at much higher speeds (Figure 10).</p>
<figure><img src="https://0mean1sigma.com/content/images/2024/11/Step2_Figure9.gif"><figcaption><span>Figure 9: Coalesced thread to element mapping</span></figcaption></figure><figure><img src="https://0mean1sigma.com/content/images/2024/11/Step2_Figure10.gif"><figcaption><span>Figure 10: Coalesced memory accesses</span></figcaption></figure><p>Modifying the kernel function is just as simple. All we need to do is change thread-to-element mapping such that thread's <em>x</em> index aligns with the column index of matrix <em><strong>C</strong></em> and <em>y</em> axis with the row index of matrix <strong><em>C</em></strong>.</p>
<pre><code>__global__ void coalesced_mat_mul_kernel(float *d_A_ptr, float *d_B_ptr, float *d_C_ptr, int C_n_rows, int C_n_cols, int A_n_cols)
{
    // Working on C[row,col]
    const int col = blockDim.x*blockIdx.x + threadIdx.x;
    const int row = blockDim.y*blockIdx.y + threadIdx.y;

    // Parallel mat mul
    if (row &lt; C_n_rows &amp;&amp; col &lt; C_n_cols)
    {
        // Value at C[row,col]
        float value = 0;
        for (int k = 0; k &lt; A_n_cols; k++)
        {
            value += d_A_ptr[row*A_n_cols + k] * d_B_ptr[k*C_n_cols + col];
        }

        // Assigning calculated value (SGEMM is C = α*(A @ B)+β*C and in this repo α=1, β=0)
        d_C_ptr[row*C_n_cols + col] = 1*value + 0*d_C_ptr[row*C_n_cols + col];
    }
}</code></pre>
<h1>Benchmark</h1>
<figure><img src="https://0mean1sigma.com/content/images/2024/09/cuBLAS_v_coalesced.png"><figcaption><span>Figure 11: cuBLAS vs Coaleased vs Uncoalesced</span></figcaption></figure><p>Figure 11 shows the GFLOPS for the coalesced and uncoalesced code against NVIDIA's SGEMM implementation. As we saw earlier that the uncoalesced version was achieving 1% of what cuBLAS can do for large matrices. With coalesced memory accesses, the kernel is at 9% of cuBLAS (which is a big jump from earlier but still not enough). It's interesting to note that our code almost on level with cuBLAS for smaller matrices (i.e. 128 x 128), and the gap widens as the matrix size increases. This is a clue that will help us improve the performance of our code further. But first, let's see what happens if we use Tensor cores instead of CUDA cores.</p>
<hr>
<h1>References</h1>
<ol><li>Coalesced matrix multiplication on a GPU</li></ol>
<figure><a href="https://github.com/tgautam03/xGeMM/blob/master/include/02_coalesced_xgemm.cuh?ref=0mean1sigma.com"><div>
<p>xGeMM/include/02_coalesced_xgemm.cuh at master · tgautam03/xGeMM</p>
<p>Accelerated General (FP32) Matrix Multiplication. Contribute to tgautam03/xGeMM development by creating an account on GitHub.</p>
</div></a><figcaption><p><span>Header File</span></p></figcaption></figure><figure><a href="https://github.com/tgautam03/xGeMM/blob/master/src/02_coalesced_xgemm.cu?ref=0mean1sigma.com"><div>
<p>xGeMM/src/02_coalesced_xgemm.cu at master · tgautam03/xGeMM</p>
<p>Accelerated General (FP32) Matrix Multiplication. Contribute to tgautam03/xGeMM development by creating an account on GitHub.</p>
</div></a><figcaption><p><span>Source File</span></p></figcaption></figure><ol start="2"><li>Benchmarking coalesced matrix multiplication on a GPU</li></ol>
<figure><a href="https://github.com/tgautam03/xGeMM/blob/master/test/02_benchmark_coalesced.cu?ref=0mean1sigma.com"><div>
<p>xGeMM/test/02_benchmark_coalesced.cu at master · tgautam03/xGeMM</p>
<p>Accelerated General (FP32) Matrix Multiplication. Contribute to tgautam03/xGeMM development by creating an account on GitHub.</p>
</div></a></figure>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via 0Mean1Sigma <br>
<br>
December 30, 2024 at 03:24PM
